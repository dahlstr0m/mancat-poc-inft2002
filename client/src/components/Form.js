// @flow

import * as React from 'react';
import { Component } from 'react-simplified';

/**
 * Renders a form label using Bootstrap styles.
 */
class FormLabel extends Component<{ children?: React.Node }> {
  render() {
    return <label className="col-form-label">{this.props.children}</label>;
  }
}

/**
 * Renders a form input using Bootstrap styles.
 */
class FormInput extends Component<{
  type: string,
  value: React.Node,
  onChange: (SyntheticEvent<HTMLInputElement>) => mixed,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, width, height, pattern
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { type, value, onChange, ...rest } = this.props;
    return (
      <input
        {...rest}
        className="form-control"
        type={this.props.type}
        value={this.props.value}
        onChange={this.props.onChange}
      />
    );
  }
}

/**
 * Renders a form textarea using Bootstrap styles.
 */
class FormTextarea extends React.Component<{
  value: React.Node,
  onChange: (SyntheticEvent<HTMLTextAreaElement>) => mixed,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, rows, cols
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { value, onChange, ...rest } = this.props;
    return <textarea {...rest} className="form-control" value={value} onChange={onChange} />;
  }
}

/**
 * Renders a form checkbox using Bootstrap styles.
 */
class FormCheckbox extends Component<{
  checked: React.Node,
  onChange: (SyntheticEvent<HTMLInputElement>) => mixed,
  labeling: React.Node,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, width, height, pattern
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { labeling, checked, onChange, ...rest } = this.props;
    return (
      <div className="custom-control custom-checkbox">
        <input
          {...rest}
          type="checkbox"
          className="custom-control-input"
          id="customCheck"
          checked={checked}
          onChange={onChange}
        />
        <label className="custom-control-label" htmlFor="customCheck">
          {labeling}
        </label>
      </div>
    );
  }
}

/**
 * Renders a form select using Bootstrap styles.
 */
class FormSelect extends Component<{
  value: React.Node,
  onChange: (SyntheticEvent<HTMLSelectElement>) => mixed,
  children?: React.Node,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, size.
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { value, onChange, children, ...rest } = this.props;
    return (
      <select {...rest} className="custom-select" value={value} onChange={onChange}>
        {children}
      </select>
    );
  }
}

/**
 * Renders a form date using Bootstrap styles.
 */
class FormDate extends Component<{
  type: string,
  value: React.Node,
  onChange: (SyntheticEvent<HTMLInputElement>) => mixed,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, width, height, pattern
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { type, value, onChange, ...rest } = this.props;
    return (
      <input
        {...rest}
        className="form-control"
        type={this.props.type}
        value={this.props.value}
        onChange={this.props.onChange}
      />
    );
  }
}

/**
 * Renders form components using Bootstrap styles.
 */
export default class Form {
  static Label = FormLabel;
  static Input = FormInput;
  static Textarea = FormTextarea;
  static Checkbox = FormCheckbox;
  static Select = FormSelect;
  static Date = FormDate;
}
