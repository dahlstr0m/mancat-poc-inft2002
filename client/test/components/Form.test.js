// @flow

import * as React from 'react';
import { shallow } from 'enzyme';

// Components
import Form from '../../src/components/Form';

// Tests
describe('Form components tests', () => {
  test('Form.Label draws correctly', () => {
    const wrapper = shallow(<Form.Label>Label</Form.Label>);

    expect(wrapper).toMatchSnapshot();
  });

  test('Form.Input draws correctly', () => {
    const wrapper = shallow(<Form.Input type="text" value="value" />);

    expect(wrapper).toMatchSnapshot();
  });

  test('Form.Textarea draws correctly', () => {
    const wrapper = shallow(<Form.Textarea value="value" />);

    expect(wrapper).toMatchSnapshot();
  });

  test('Form.Checkbox draws correctly', () => {
    const wrapper = shallow(<Form.Checkbox checked labeling="Check-label" />);

    expect(wrapper).toMatchSnapshot();
  });

  test('Form.Select draws correctly', () => {
    const wrapper = shallow(<Form.Select value={2}></Form.Select>);

    expect(wrapper).toMatchSnapshot();
  });
});
