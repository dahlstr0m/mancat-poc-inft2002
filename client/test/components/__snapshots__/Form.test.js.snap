// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form components tests Form.Checkbox draws correctly 1`] = `
<div
  className="custom-control custom-checkbox"
>
  <input
    checked={true}
    className="custom-control-input"
    id="customCheck"
    type="checkbox"
  />
  <label
    className="custom-control-label"
    htmlFor="customCheck"
  >
    Check-label
  </label>
</div>
`;

exports[`Form components tests Form.Input draws correctly 1`] = `
<input
  className="form-control"
  type="text"
  value="value"
/>
`;

exports[`Form components tests Form.Label draws correctly 1`] = `
<label
  className="col-form-label"
>
  Label
</label>
`;

exports[`Form components tests Form.Select draws correctly 1`] = `
<select
  className="custom-select"
  value={2}
/>
`;

exports[`Form components tests Form.Textarea draws correctly 1`] = `
<textarea
  className="form-control"
  value="value"
/>
`;
